sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -56,117 +52,247 @@ case class SortMergeJoin(
-              // for inner join, we need to filter those null keys
-              stop = comparing == 0 && !leftKey.anyNull
-              if (comparing > 0 || rightKey.anyNull) {
-                fetchRight()
-              } else if (comparing < 0 || leftKey.anyNull) {
-                fetchLeft()
-              }
-            }
-            rightMatches = new CompactBuffer[InternalRow]()
-            if (stop) {
-              stop = false
-              // iterate the right side to buffer all rows that matches
-              // as the records should be ordered, exit when we meet the first that not match
-              while (!stop && rightElement != null) {
-                rightMatches += rightElement
-                fetchRight()
-                stop = keyOrdering.compare(leftKey, rightKey) != 0
-              }
-              if (rightMatches.size > 0) {
-                rightPosition = 0
-                matchKey = leftKey
-              }
-            }
+  /**
+   * Advances the streamed input iterator and buffers all rows from the build input with matching
+   * keys.
+   * @return true if the streamed iterator returned a row, false otherwise. If this returns true,
+   *         then [getStreamedRow and [[getBuildMatches]] can be called to produce the outer
+   *         join results.
+   */
+  final def findNextOuterJoinRows(): Boolean = {
+    if (advancedStreamed()) {
+      if (streamedRowKey.anyNull) {
+        // Since at least one join column is null, the streamed row has no matches.
+        matchJoinKey = null
+        buildMatches = null
+      } else if (matchJoinKey != null && keyOrdering.compare(streamedRowKey, matchJoinKey) == 0) {
+        // Matches the current group, so do nothing.
+      } else {
+        // The streamed row does not match the current group.
+        matchJoinKey = null
+        buildMatches = null
+        if (buildRow != null) {
+          // The build iterator could still contain matching rows, so we'll need to walk through it
+          // until we either find matches or pass where they would be found.
+          var comp =
+            if (buildRowKey.anyNull) 1 else keyOrdering.compare(streamedRowKey, buildRowKey)
+          while (comp > 0 && advanceBuild()) {
+            comp = if (buildRowKey.anyNull) 1 else keyOrdering.compare(streamedRowKey, buildRowKey)
+          }
+          if (comp == 0) {
+            // We have found matches, so buffer them (this updates matchJoinKey)
+            bufferMatchingBuildRows()
+          } else {
+            // We have overshot the position where the row would be found, hence no matches.
           }
-          rightMatches != null && rightMatches.size > 0
         }
       }
+      // If there is a streamed input, then we always return true
+      true
+    } else {
+      // End of streamed input, hence no more results.
+      false
     }
   }
+
+  def getStreamedRow: InternalRow = streamedRow
```

Comment:
Move these two methods up; add a higher-level summary how how consumers use this in the top Scaladoc.